Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp

```diff
@@ -130,6 +131,50 @@ std::optional<CString> XDGDBusProxy::accessibilityProxy(const char* baseDirector
 #endif
 }
 
+static void waitUntilSyncedOrDie(GSubprocess* subprocess, int syncFd)
+{
+    GRefPtr<GMainContext> context = adoptGRef(g_main_context_new());
+    g_main_context_push_thread_default(context.get());
+
+    struct AsyncData {
+        int opsFinished = 0;
+        GUniqueOutPtr<GError> error;
+    } data;
+    GRefPtr<GCancellable> cancellable = adoptGRef(g_cancellable_new());
+
+    g_subprocess_wait_check_async(subprocess, cancellable.get(), +[](GObject* source, GAsyncResult* result, gpointer data) {
+        GUniqueOutPtr<GError> error;
+        auto* asyncData = static_cast<AsyncData*>(data);
+        g_subprocess_wait_check_finish(G_SUBPROCESS(source), result, &error.outPtr());
+        if (error && !g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
+            g_propagate_error(&asyncData->error.outPtr(), error.release());
+        asyncData->opsFinished++;
+    }, &data);
+
+    GRefPtr<GInputStream> inputStream = adoptGRef(g_unix_input_stream_new(syncFd, FALSE));
+    char out;
+    g_input_stream_read_async(inputStream.get(), &out, 1, G_PRIORITY_DEFAULT, cancellable.get(), +[](GObject* source, GAsyncResult* result, gpointer data) {
+        GUniqueOutPtr<GError> error;
+        auto* asyncData = static_cast<AsyncData*>(data);
+        g_input_stream_read_finish(G_INPUT_STREAM(source), result, &error.outPtr());
+        if (error && !g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
+            g_propagate_error(&asyncData->error.outPtr(), error.release());
+        asyncData->opsFinished++;
+    }, &data);
+
+    while (data.opsFinished != 2) {
+        g_main_context_iteration(context.get(), TRUE);
+        if (data.opsFinished == 1 && cancellable) {
```

Comment:
Right.

I can add a comment.

The check `&& cancellable` here is actually not needed (because `g_cancellable_cancel()` is NULL-safe), but I think it improves readability.